### PR TITLE
fix monet unet

### PR DESCRIPTION
**Pull request by hnyu**
The current UNet has some issue that the channel numbers of the downsampling path and those of the upsampling path are not always symmetric. This causes an issue for creating skip connections. For example, for a filter list of (32, 64, 64, 128, 128, 128), the result channels are:

```python
DOWN OUTPUT SHAPE:  torch.Size([2, 15, 96, 128])
DOWN OUTPUT SHAPE:  torch.Size([2, 32, 96, 128])
DOWN OUTPUT SHAPE:  torch.Size([2, 64, 48, 64])
DOWN OUTPUT SHAPE:  torch.Size([2, 64, 24, 32])
DOWN OUTPUT SHAPE:  torch.Size([2, 128, 12, 16])
DOWN OUTPUT SHAPE:  torch.Size([2, 128, 6, 8])
UP OUTPUT SHAPE:  torch.Size([2, 128, 3, 4])
UP OUTPUT SHAPE:  torch.Size([2, 128, 6, 8])
UP OUTPUT SHAPE:  torch.Size([2, 128, 12, 16])
UP OUTPUT SHAPE:  torch.Size([2, 128, 24, 32])
```

Then 128 is concat with 64, the network crashes. 

This PR fixes this asymmetry issue.

### CODE REVIEW

File: alf/algorithms/monet_algorithm.py [hunk 1]
```diff
@@ -90,42 +90,43 @@ def __init__(self,
...
-            fc_layer_params=nonskip_fc_layers,
-            last_layer_size=last_skip_tensor_spec.numel,
-            last_activation=torch.relu_)
+            input_tensor_spec=alf.TensorSpec((last_skip_tensor_spec.numel, )),
+            fc_layer_params=nonskip_fc_layers)
 
-        self._reshape = alf.layers.Reshape(*last_skip_tensor_spec.shape)
+        self._reshape = torch.nn.Sequential(
+            alf.layers.FC(nonskip_fc_layers[-1], last_skip_tensor_spec.numel,
+                          torch.relu_),
```
Comment (emailweixu) [inline alf/algorithms/monet_algorithm.py:101]: This can be part of nonskip_mlp. And should update document of nonskip_fc_layers about the additional FC layer.

Comment (hnyu) [reply alf/algorithms/monet_algorithm.py:101]: > This can be part of nonskip_mlp. And should update document of nonskip_fc_layers about the additional FC layer.

The reason why I chose to put this additional FC layer in `self._reshape` instead of `self._nonskip_mlp` is that sometimes we want to perform extra operations on the output of `self._nonskip_mlp` (as in the vonet case). This output can be obtained by calling `self.encode()`. If the additional FC is in `self._nonskip_mlp` then the encoding dim might be large depending on `last_skip_tensor_spec.numel` and is not straightforwardly controllable.

Comment (emailweixu) [reply alf/algorithms/monet_algorithm.py:101]: Should still comment about this additional fc in doc-string of the class.

Comment (hnyu) [reply alf/algorithms/monet_algorithm.py:101]: > Should still comment about this additional fc in doc-string of the class.

added

**Review by emailweixu — APPROVED** [review]
